# 1141_社會資料分析 第一次期中考
學生：劉晏成



In [1]:
from load import load_sav
import pandas as pd
import numpy as np
import statistic as my_stat

In [2]:
df = load_sav("../data/midterm_1.sav")
df.loc[:, 'age'] = 2018 - (df['a2_1'] + 1911)

In [3]:
variable_value_labels = df.attrs['variable_value_labels']
column_names = df.attrs['column_names']
column_names_to_labels = df.attrs['column_names_to_labels']

In [4]:
# variable_value_labels['g4a']

## Q1. 
這份問卷調查執行於 2018 年，請利用「出生年」計算，創造出「年齡」變項。接著，請列出（1）性別、 （2）年齡、 （3）教育程度、 （4）家庭收入等變項的次數分布。

In [5]:
df_1 = df[['a1', 'age', 'b3', 'a8']].copy().rename(columns = { 'a1': 'sex', 'a8': 'education', 'b3': 'family_income' })

### Ans 1: 性別次數分布

In [6]:
df_1['sex'].value_counts().sort_index()

sex
1.0    418
2.0    450
Name: count, dtype: int64

### Ans 2: 年齡次數分布

In [7]:
with pd.option_context("display.max_rows", None,
                       "display.max_colwidth", 200):
    display(df_1['age'].value_counts().sort_index())

age
18.0     4
19.0     8
20.0     7
21.0    17
22.0    15
23.0    17
24.0    29
25.0    14
26.0    17
27.0    19
28.0    22
29.0    15
30.0    11
31.0    19
32.0    20
33.0    23
34.0    16
35.0    21
36.0    26
37.0    21
38.0    24
39.0    18
40.0    17
41.0    23
42.0    30
43.0    16
44.0    20
45.0    16
46.0    15
47.0    14
48.0    17
49.0    11
50.0    13
51.0    14
52.0    15
53.0    22
54.0    24
55.0    20
56.0    20
57.0    20
58.0    11
59.0    13
60.0    11
61.0    13
62.0    13
63.0    14
64.0     8
65.0    14
66.0     8
67.0     8
68.0     8
69.0     7
70.0     5
71.0     3
72.0     1
73.0     6
74.0     1
75.0     2
76.0     1
77.0     4
78.0     3
80.0     2
81.0     1
83.0     1
Name: count, dtype: int64

### Ans 3: 教育程度分布

In [8]:
df_1['education'].value_counts().sort_index()

education
1.0       4
2.0       1
3.0      55
4.0      91
5.0       1
6.0      31
7.0      96
8.0     161
9.0       2
10.0     41
11.0     49
12.0      6
13.0      1
14.0      5
15.0      3
16.0      4
17.0      2
18.0     60
19.0    210
20.0     41
21.0      4
Name: count, dtype: int64

### Ans 4: 家庭收入分布

In [9]:
df_1['family_income'].value_counts().sort_index()

family_income
0.0      21
1.0      27
2.0     126
3.0     137
4.0     166
5.0     101
6.0     108
7.0      46
8.0      40
9.0      18
10.0     15
11.0     52
Name: count, dtype: int64

## Q2. 
請做下列計算，並說明結果：  
（1）性別的眾數（指出是哪個性別）；  
（2）平均年齡；  
（3）教育程度的眾數； （指出是哪個教育程度；如： 「小學」 ）；  
（4）家庭收入的中位數（指出是哪個收入層級；如「2 萬元以上，未滿 4 萬元」） 。

In [10]:
df_2 = df[['a1', 'age', 'b3', 'a8']].copy().rename(columns = { 'a1': 'sex', 'a8': 'education', 'b3': 'family_income' })

### Ans 1: 性別的眾數

再次得出性別次數分布

In [11]:
df_2['sex'].value_counts().sort_index()

sex
1.0    418
2.0    450
Name: count, dtype: int64

眾數為450，為女性。

### Ans 2: 平均年齡

計算出年齡的描述性統計

In [12]:
df_2['age'].describe()

count    868.000000
mean      43.267281
std       14.500049
min       18.000000
25%       32.000000
50%       42.000000
75%       55.000000
max       83.000000
Name: age, dtype: float64

平均年齡為43.27歲

### Ans 3: 教育程度的眾數； （指出是哪個教育程度；如： 「小學」 ）

計算出教育程度的眾數

In [13]:
f"mod: {df_2.loc[:, 'education'].mode()[0]}"

'mod: 19.0'

眾數為19 ，為大學

### Ans 4: 家庭收入的中位數（指出是哪個收入層級；如「2 萬元以上，未滿 4 萬元」）

計算出家庭收入的描述性統計

In [14]:
df_2['family_income'].describe()

count    857.000000
mean       4.705951
std        2.629341
min        0.000000
25%        3.000000
50%        4.000000
75%        6.000000
max       11.000000
Name: family_income, dtype: float64

中位數為4，為"6萬元以上，未滿8萬元"

## Q3.
將「教育程度」重新編碼成新變項「大專以上學歷」（變項名 college_y） 。「無」到「士官學校」編成0（數值標籤設成「大專以下」） ； 「五專」到「博士」編成 1（「大專以上」） 。然後列出新變項「大專以上學歷」的次數分布。

In [15]:
df_3 = df[['a8']].copy().rename(columns = { 'a8': 'education' })

In [16]:
df_3.attrs['variable_value_labels']['a8']

{1.0: '無',
 2.0: '自修',
 3.0: '小學肄業',
 4.0: '國(初)中',
 5.0: '初職',
 6.0: '高中普通科',
 7.0: '高中職業科',
 8.0: '高職',
 9.0: '士官學校',
 10.0: '五專',
 11.0: '二專',
 12.0: '三專',
 13.0: '軍警專修班',
 14.0: '軍警專科班',
 15.0: '空中行專',
 16.0: '空中大學',
 17.0: '軍警官學校',
 18.0: '技術學院、科大',
 19.0: '大學',
 20.0: '碩士',
 21.0: '博士',
 22.0: '其他',
 997.0: '拒答',
 998.0: '漏答',
 999.0: '不適用'}

依照指示編碼，並將22、997、998、999 設為遺漏值

In [17]:
import numpy as np

df_3_recoded = (
    df_3.assign(
        college_y=np.select(
            [df_3["education"].between(1, 9), df_3["education"].between(10, 21)],
            [0, 1],
            default=np.nan
        )
    )
)
df_3_recoded

education  college_y
0         18.0        1.0
1         18.0        1.0
2         18.0        1.0
3          8.0        0.0
4          6.0        0.0
..         ...        ...
863        3.0        0.0
864        8.0        0.0
865        3.0        0.0
866        8.0        0.0
867        4.0        0.0

[868 rows x 2 columns]

### Ans
計算次數分布

In [18]:
with pd.option_context("display.max_rows", None,
                       "display.max_colwidth", 200):
    display(df_3_recoded['college_y'].value_counts().sort_index())

college_y
0.0    442
1.0    426
Name: count, dtype: int64

## Q4. 
將「家庭收入」重新編碼成新變項「收入等級」（變項名 f_income_3gp） 。 「無收入」到未滿 6 萬元編
成 1（數值標籤設成「低」）；6 萬元以上到未滿 10 萬元編成 2（「中」）；10 萬元以上編成 3（「高」）。
然後列出新變項「收入等級」的次數分布。

In [19]:
df_4 = df[['b3']].copy().rename(columns = { 'b3': 'family_income' })

依照題目進行編碼

In [20]:
df_4_recoded = df_4.assign(
    f_income_3gp_num=np.select(
        [df_4.family_income.between(0, 3), df_4.family_income.between(4, 5), df_4.family_income.between(6, 11)],
        [0, 1, 2],
        default=np.nan
    )
)

df_4_recoded["f_income_3gp"] = df_4_recoded["f_income_3gp_num"].map({0: "低", 1: "中", 2: "高"})
df_4_recoded

family_income  f_income_3gp_num f_income_3gp
0              4.0               1.0            中
1              3.0               0.0            低
2              4.0               1.0            中
3              2.0               0.0            低
4              2.0               0.0            低
..             ...               ...          ...
863           11.0               2.0            高
864           11.0               2.0            高
865            0.0               0.0            低
866            3.0               0.0            低
867           11.0               2.0            高

[868 rows x 3 columns]

### Ans
計算次數分布

In [21]:
display(df_4_recoded['f_income_3gp'].value_counts(dropna=False).sort_index())

f_income_3gp
中      267
低      311
高      279
NaN     11
Name: count, dtype: int64

## Q5. 
按下列步驟，創造「政治自由態度」變項。  
（1）將「拒答、漏答、不適用」等回應設成遺漏值；  
（2）檢查 g4a 到 g4i 這 9 個變項分布，如有其他不合理值，也設成遺漏值；  
（3）將 g4a、g4c、g4e、g4f、g4h 做反向編碼（將 1 改成 4，2 改成 3，依此類推） ；  
（4）將反向編碼後的 5 個變項，和其他 4 個變項的得分加總，將總分變項設成：G4_all，並加上標籤「政治自由態度」。    
（5）計算並報告「政治自由態度」變項的平均數和標準差。  

In [22]:
df_5 = df.loc[:, "g4a":"g4i"].copy()

觀察九個欄位的次數分布

In [23]:
out = df_5.apply(pd.Series.value_counts, dropna=False)
print(out)

         g4a    g4b    g4c    g4d    g4e    g4f    g4g    g4h  g4i
1.0    164.0   18.0   55.0   33.0  129.0  209.0   10.0   63.0   24
2.0    474.0   91.0  339.0  258.0  487.0  462.0   65.0  229.0  267
3.0    185.0  550.0  350.0  459.0  210.0  158.0  502.0  285.0  445
4.0     42.0  207.0  121.0  114.0   39.0   35.0  288.0  288.0  127
997.0    2.0    1.0    2.0    2.0    2.0    2.0    2.0    2.0    3
998.0    1.0    1.0    1.0    2.0    1.0    2.0    1.0    1.0    1
999.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    1


將997.0（'拒答'）、 998.0（'漏答'）、 999.0（'不適用'）設為遺漏值。再觀察次數分布作為檢驗

In [24]:
df_5_cleared = df_5.loc[:, "g4a":"g4i"].where(df_5.loc[:, "g4a":"g4i"].isin([1, 2, 3, 4]))
out = df_5_cleared.apply(pd.Series.value_counts, dropna=False)
print(out)

     g4a  g4b  g4c  g4d  g4e  g4f  g4g  g4h  g4i
1.0  164   18   55   33  129  209   10   63   24
2.0  474   91  339  258  487  462   65  229  267
3.0  185  550  350  459  210  158  502  285  445
4.0   42  207  121  114   39   35  288  288  127
NaN    3    2    3    4    3    4    3    3    5


將 g4a、g4c、g4e、g4f、g4h 做反向編碼（將 1 改成 4，2 改成 3，依此類推）

In [25]:
# Columns to reverse-code and mapping
rev_cols = ["g4a", "g4c", "g4e", "g4f", "g4h"]
mapping = {1: 4, 2: 3, 3: 2, 4: 1}

# Make a new dataframe with reverse-coded values
df_5_reverse_recoded = df_5_cleared.copy()
df_5_reverse_recoded[rev_cols] = df_5_reverse_recoded[rev_cols].replace(mapping)
df_5_reverse_recoded

g4a  g4b  g4c  g4d  g4e  g4f  g4g  g4h  g4i
0    3.0  3.0  2.0  3.0  3.0  3.0  3.0  2.0  3.0
1    3.0  3.0  3.0  3.0  3.0  2.0  3.0  2.0  3.0
2    3.0  3.0  2.0  3.0  3.0  3.0  3.0  1.0  2.0
3    1.0  4.0  1.0  4.0  1.0  1.0  4.0  1.0  4.0
4    3.0  3.0  2.0  3.0  3.0  3.0  3.0  2.0  2.0
..   ...  ...  ...  ...  ...  ...  ...  ...  ...
863  4.0  4.0  2.0  3.0  1.0  4.0  4.0  1.0  4.0
864  3.0  3.0  2.0  2.0  3.0  3.0  3.0  3.0  3.0
865  4.0  4.0  1.0  2.0  2.0  4.0  4.0  3.0  4.0
866  4.0  4.0  3.0  3.0  3.0  3.0  4.0  1.0  3.0
867  3.0  3.0  3.0  4.0  2.0  2.0  4.0  1.0  3.0

[868 rows x 9 columns]

將反向編碼後的 5 個變項，和其他 4 個變項的得分加總，將總分變項設成：G4_all，並加上標籤「政治自由態度」。  
檢視表格

In [26]:
df_5_final = df_5_reverse_recoded.copy()
df_5_final["G4_all"] = df_5_final.sum(axis=1)
df_5_final

g4a  g4b  g4c  g4d  g4e  g4f  g4g  g4h  g4i  G4_all
0    3.0  3.0  2.0  3.0  3.0  3.0  3.0  2.0  3.0    25.0
1    3.0  3.0  3.0  3.0  3.0  2.0  3.0  2.0  3.0    25.0
2    3.0  3.0  2.0  3.0  3.0  3.0  3.0  1.0  2.0    23.0
3    1.0  4.0  1.0  4.0  1.0  1.0  4.0  1.0  4.0    21.0
4    3.0  3.0  2.0  3.0  3.0  3.0  3.0  2.0  2.0    24.0
..   ...  ...  ...  ...  ...  ...  ...  ...  ...     ...
863  4.0  4.0  2.0  3.0  1.0  4.0  4.0  1.0  4.0    27.0
864  3.0  3.0  2.0  2.0  3.0  3.0  3.0  3.0  3.0    25.0
865  4.0  4.0  1.0  2.0  2.0  4.0  4.0  3.0  4.0    28.0
866  4.0  4.0  3.0  3.0  3.0  3.0  4.0  1.0  3.0    28.0
867  3.0  3.0  3.0  4.0  2.0  2.0  4.0  1.0  3.0    25.0

[868 rows x 10 columns]

計算並報告「政治自由態度」變項的平均數和標準差

計算「政治自由態度」的敘述性統計

In [27]:
df_5_final['G4_all'].describe()

count    868.00000
mean      24.89977
std        3.36175
min        0.00000
25%       23.00000
50%       25.00000
75%       27.00000
max       36.00000
Name: G4_all, dtype: float64

### Ans
政治自由態度的平均數為24.90、標準差為3.36

## Q6. 以「單一樣本 T 檢定」檢視受訪者的「政治自由態度」平均分數是否為 24。

In [28]:
df_6 = df_5_final.copy()

以樣本做母體平均數檢定

* $H0: \mu = 24$
* $H1: \mu \neq 24$

並且以$\alpha$ < 0.05 為檢定判準

In [29]:
from scipy import stats

mu_0 = 24

# two-sided for for two-sided test; less for case you want to check if the sample mean is less than population mean
t_statistic, p_value = stats.ttest_1samp(df_6['G4_all'], mu_0, alternative='two-sided')

pd.DataFrame({
    'Values': [f'{t_statistic:.2f}', f'{p_value:.15e}']
}, index=['statstic', 'P Value'])

Values
statstic                   7.89
P Value   9.403443175267619e-15

### Ans
觀察P Value < 0.05 ，故拒絕虛無假設，母體平均分數不為24。

## Q7. 以「獨立樣本 T 檢定」比較「男性」和「女性」的「政治自由態度」平均數是否有顯著差異。

預覽二欄位

In [30]:
# pick columns
col1 = df[['a1']].copy().rename(columns = { 'a1': 'sex' })['sex']
col2 = df_5_final['G4_all']
df_7 = pd.concat([col1.reset_index(drop=True),
                  col2.reset_index(drop=True)], axis=1)
df_7

sex  G4_all
0    2.0    25.0
1    1.0    25.0
2    2.0    23.0
3    1.0    21.0
4    2.0    24.0
..   ...     ...
863  2.0    27.0
864  2.0    25.0
865  2.0    28.0
866  1.0    28.0
867  1.0    25.0

[868 rows x 2 columns]

In [31]:
df_7_male = df_7.loc[df_7['sex'] == 1.0, 'G4_all']
df_7_female = df_7.loc[df_7['sex'] == 2.0, 'G4_all']

考察兩資料的性質為獨立樣本

將男女兩組資料分組後，進行Levene's test 以及接續之T Test

In [32]:
from statistic import t_test

t_test.compare_independent_samples(df_7_male, df_7_female) 

Statistic   p-value
Levene's Test for Equal Variances     0.985017  0.321241
t-test (Equal Variance Assumed)       0.785801  0.432199
t-test (Equal Variance Not Assumed)   0.782183  0.434330

### Ans
**Levene's Test**  
H0: 兩母群體變異數相同  
H1: 兩母群體變異數不相同  
$\alpha = 0.05$

首先檢視表格中Levene's Test 的P Value > 0.05，接受Levenu's test 的H0（兩母群體變異數相同）。

接下觀察母群體變異數相同的T Test 結果

設定：
男性、女性政治自由態度平均數為 $\mu_1$、$\mu_2$  
$H_0: \mu_1 = \mu_2$  
$H_1: \mu_1 \neq \mu_2$  
$\alpha = 0.05$

由此設定進一步檢視假設變異數相同的分析結果，P Value > 0.05，接受虛無假設：不同性別與政治自由態度**無顯著差異**。

## Q8. 以「獨立樣本 T 檢定」比較「大專以下」和「大專以上」學歷者的「政治自由態度」平均數是否有顯著差異。

In [33]:
# pick columns
col_8_1 = df_3_recoded['college_y']
col_8_2 = df_5_final['G4_all']
df_8 = pd.concat([col_8_1.reset_index(drop=True),
                  col_8_2.reset_index(drop=True)], axis=1)
df_8

college_y  G4_all
0          1.0    25.0
1          1.0    25.0
2          1.0    23.0
3          0.0    21.0
4          0.0    24.0
..         ...     ...
863        0.0    27.0
864        0.0    25.0
865        0.0    28.0
866        0.0    28.0
867        0.0    25.0

[868 rows x 2 columns]

In [34]:
df_8_0 = df_8.loc[df_8['college_y'] == 0, 'G4_all']
df_8_1 = df_8.loc[df_8['college_y'] == 1, 'G4_all']

考察兩資料的性質為獨立樣本

將兩個分類資料分組後，進行Levene's test 以及接續之T Test

In [35]:
from statistic import t_test

t_test.compare_independent_samples(df_8_0, df_8_1) 

Statistic       p-value
Levene's Test for Equal Variances     4.052579  4.441295e-02
t-test (Equal Variance Assumed)      -6.052369  2.123467e-09
t-test (Equal Variance Not Assumed)  -6.054851  2.092464e-09

### Ans
**Levene's Test**  
H0: 兩母群體變異數相同  
H1: 兩母群體變異數不相同  
$\alpha = 0.05$

首先檢視表格中Levene's Test 的P Value < 0.05，拒絕Levenu's test 的H0（兩母群體變異數不相同）。

接下觀察母群體變異數相同的T Test 結果

設定：
大專以上、大專以下學歷政治自由態度平均數為 $\mu_1$、$\mu_2$  
$H_0: \mu_1 = \mu_2$  
$H_1: \mu_1 \neq \mu_2$  
$\alpha = 0.05$

由此設定進一步檢視不假設變異數相同的分析結果，P Value < 0.05，拒絕虛無假設：是否擁有大專學歷與政治自由態度**有顯著差異**。

## Q9. 以變異數分析方法（包括整體檢定與事後比較），檢視不同「收入等級」者的「政治自由態度」平均分數是否有顯著差異。

In [36]:
# pick columns
col_9_1 = df_4_recoded["f_income_3gp"]
col_9_2 = df_5_final['G4_all']
df_9 = pd.concat([col_9_1.reset_index(drop=True),
                  col_9_2.reset_index(drop=True)], axis=1)

檢視遺漏值狀況

In [37]:
from statistic.missing_value import check_missing_value

check_missing_value(df_9)

# of Missing Values
f_income_3gp                   11
G4_all                          0

發現有缺漏值，預處理f_income_3gp 為缺漏值的資料

In [38]:
df_9_cleared = df_9.dropna()
check_missing_value(df_9_cleared)

# of Missing Values
f_income_3gp                    0
G4_all                          0

執行ANOVA  
設定：不同收入等級的政治自由度態度平均數分別為 $\mu_1$、$\mu_2$、$\mu_3$  
$H_0: \mu_1 = \mu_2 = \mu_3$  
$H_1: \mu_k 不完全相等$  
$\alpha = 0.05$


In [39]:
from statistic.anova import perform_anova


perform_anova(df=df_9_cleared, val_col='G4_all', group_col='f_income_3gp')

sum_sq     df             F    PR(>F)
Intercept        168026.981273    1.0  14968.694231  0.000000
C(f_income_3gp)     119.266944    2.0      5.312452  0.005094
Residual           9586.343324  854.0           NaN       NaN

依據ANOVA 結果，f_income_3gp 的P value < 0.05 ，拒絕虛無假設。結果顯示**組別間存在差異**

接下來執行Levene's Test 來判斷均值性狀況，以執行事後比較

In [40]:
from scipy.stats import levene

groups = {name: group['G4_all'].values for name, group in df_9_cleared.groupby('f_income_3gp')}

levene_stat, levene_p = levene(*groups.values())
pd.DataFrame({ 'Values': [f"{levene_stat:.2f}", levene_p] }, index = ['Levene Stat', 'P Value'])

Values
Levene Stat      1.16
P Value      0.313178

結果顯示P Value > 0.05：變異數同質假定未違反  
選擇HSD法

In [41]:
from statistic.anova import perform_post_hoc

perform_post_hoc(df_9_cleared, levene_p = levene_p, val_col='G4_all', group_col='f_income_3gp')

A  B    mean(A)    mean(B)      diff        se         T   p-tukey  \
0  中  低  25.086142  24.430868  0.655274  0.279528  2.344215  0.050453   
1  中  高  25.086142  25.286738 -0.200596  0.286838 -0.699336  0.763932   
2  低  高  24.430868  25.286738 -0.855870  0.276275 -3.097895  0.005705   

     hedges  
0  0.198268  
1 -0.059900  
2 -0.251109

### Ans
由Anova 以及事後多重比較結果可得知：
1. 三組收入等級組別的平均收入不完全相同
2. 收入等級與政治自由態度關係為：中 > 低（未達顯著）、高 > 中（未達顯著）、高 > 低（達顯著）
3. 承上，統計上，可知高 > 低，其餘未達顯著標準

## Q10. 以「年齡」作為共變量，再次檢視不同「收入等級」者的「政治自由態度」平均分數是否有顯著差異。

In [42]:
# pick columns
col_10_1 = df_4_recoded["f_income_3gp"]
col_10_2 = df['age']
col_10_3 = df_5_final['G4_all']

df_10 = pd.concat([col_10_1.reset_index(drop=True),
                  col_10_2.reset_index(drop=True), col_10_3.reset_index(drop=True)], axis=1)

df_10

f_income_3gp   age  G4_all
0              中  30.0    25.0
1              低  31.0    25.0
2              中  33.0    23.0
3              低  24.0    21.0
4              低  59.0    24.0
..           ...   ...     ...
863            高  58.0    27.0
864            高  38.0    25.0
865            低  68.0    28.0
866            低  39.0    28.0
867            高  65.0    25.0

[868 rows x 3 columns]

檢查遺漏值狀況

In [43]:
from statistic.missing_value import check_missing_value

check_missing_value(df_10)

# of Missing Values
f_income_3gp                   11
age                             0
G4_all                          0

In [44]:
df_10_cleared = df_10.dropna()
check_missing_value(df_10_cleared)

# of Missing Values
f_income_3gp                    0
age                             0
G4_all                          0

基於ANCOVA 的兩個假設做假設檢驗
1. **共變量與因變量**需為線性關係
2. **共變量與自變量（組別）間** 不得有交互作用

首先檢驗共變數與因變數是否為線性關係

In [45]:
from statistic.anova import check_linearity

check_linearity(df_10_cleared, val_col='G4_all', covar_col='age')

sum_sq     df            F   PR(>F)
Intercept  57509.371598    1.0  5115.331554  0.00000
age           93.229879    1.0     8.292592  0.00408
Residual    9612.380390  855.0          NaN      NaN

age 的P Value < 0.05 可知教育年數與收入符合線性關係。

往下繼續檢查共變數與自變數的互動關係：檢驗迴歸同質性

In [46]:
from statistic.anova import check_interaction

check_interaction(df_10_cleared, val_col='G4_all', group_col='f_income_3gp', covar_col='age')

sum_sq     df            F         PR(>F)
Intercept            16210.872329    1.0  1447.571715  8.347990e-186
C(f_income_3gp)         18.289744    2.0     0.816604   4.422760e-01
age                     25.684484    1.0     2.293531   1.302854e-01
C(f_income_3gp):age      3.069263    2.0     0.137037   8.719572e-01
Residual              9530.064873  851.0          NaN            NaN

由agegroup: eduy 的P value > 0.05 可知收入等級分組與年齡交互作用不顯著，可繼續進行ANCOVA。

進行ANCOVA

In [47]:
from statistic.anova import perform_ancova
perform_ancova(df=df_10_cleared, val_col='G4_all', group_col='f_income_3gp', covar_col='age')

df       sum_sq    mean_sq         F    PR(>F)
C(f_income_3gp)    2.0   119.266944  59.633472  5.335848  0.004978
age                1.0    53.209188  53.209188  4.761020  0.029384
Residual         853.0  9533.134136  11.176007       NaN       NaN

### Ans
f_income_3gp 的P value < 0.05

控制年齡後，收入等級分組與政治自由態度**仍有差異**